ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/SandboxLedgerBackend.scala

```diff
@@ -31,19 +31,37 @@ class SandboxLedgerBackend(ledger: Ledger)(implicit mat: Materializer) extends L
       ledger.publishTransaction(submitted)
     }
 
+    private[this] def canSeeContract(
+        submitter: Party,
+        ac: ActiveContracts.ActiveContract): Boolean = {
+      // ^ only parties disclosed or divulged to can lookup; see https://github.com/digital-asset/daml/issues/10
+      // and https://github.com/digital-asset/daml/issues/751 .
+      Ref.Party fromString submitter exists (p => ac.witnesses(p) || ac.divulgences(p))
+    }
+
     override def lookupActiveContract(submitter: Party, contractId: Value.AbsoluteContractId)
       : Future[Option[Value.ContractInst[TxValue[Value.AbsoluteContractId]]]] =
       ledger
         .lookupContract(contractId)
         .map(_.collect {
-          case ac
-              if Ref.Party fromString submitter exists (p => ac.witnesses(p) || ac.divulgences(p)) =>
-            // ^ only parties disclosed or divulged to can lookup; see https://github.com/digital-asset/daml/issues/10
-            ac.contract
+          case ac if canSeeContract(submitter, ac) => ac.contract
         })(DirectExecutionContext)
 
-    override def lookupContractKey(key: Node.GlobalKey): Future[Option[Value.AbsoluteContractId]] =
-      ledger.lookupKey(key)
+    override def lookupContractKey(
+        submitter: Party,
+        key: Node.GlobalKey): Future[Option[Value.AbsoluteContractId]] = {
+      implicit val ec: ExecutionContext = DirectExecutionContext
+      ledger.lookupKey(key).flatMap {
+        // note that we need to check visibility for keys, too, otherwise we leak the existence of a non-divulged
+        // contract if we return `Some`.
+        case None => Future.successful(None)
+        case Some(cid) =>
+          ledger.lookupContract(cid) map {
+            case Some(ac) if canSeeContract(submitter, ac) => Some(cid)
```

Comment:
Because guards break exhaustiveness checking:

```suggestion
            _ flatMap (ac => if (canSeeContract(submitter, ac))  Some(cid)
```

Reply:
done